ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Run.hs

```diff
@@ -0,0 +1,311 @@
+            Nothing ->
+              pure ()
+          throwIO exn
+        Right res' -> pure res'
+  pure chainSyncException
+
+-- | Start the BlockFetch client, using the supplied 'FetchClientRegistry' to
+-- register it for synchronization with the ChainSync client.
+startBlockFetchConnectionThread ::
+  (IOLike m, MonadTime m) =>
+  ResourceRegistry m ->
+  FetchClientRegistry PeerId (Header TestBlock) TestBlock m ->
+  ControlMessageSTM m ->
+  SharedResources m ->
+  m ()
+startBlockFetchConnectionThread registry fetchClientRegistry controlMsgSTM SharedResources {srPeerId, srBlockTree, srCurrentState} =
+  void $ forkLinkedThread registry ("BlockFetchClient" <> condense srPeerId) $
+    PeerSimulator.BlockFetch.runBlockFetchClient srPeerId fetchClientRegistry controlMsgSTM getCurrentChain
+  where
+    getCurrentChain = atomically $ do
+      nodeState <- readTVar srCurrentState
+      case nodeState of
+        Nothing -> retry
+        Just aps -> do
+          let PointSchedule.BlockPoint b = PointSchedule.block aps
+          case BT.findFragment (blockPoint b) srBlockTree of
+            Just f  -> pure f
+            Nothing -> error "block tip is not in the block tree"
+
+-- | The 'Tick' contains a state update for a specific peer.
+-- If the peer has not terminated by protocol rules, this will update its TMVar
+-- with the new state, thereby unblocking the handler that's currently waiting
+-- for new instructions.
+dispatchTick ::
+  IOLike m =>
+  Tracer m String ->
+  Map PeerId (PeerResources m) ->
+  Tick ->
+  m ()
+dispatchTick tracer peers Tick {active = Peer pid state} =
+  case peers Map.!? pid of
+    Just PeerResources {prChainSync = ChainSyncResources {csrNextState}} -> do
+      trace $ "Writing state " ++ condense state
+      atomically (tryPutTMVar csrNextState state) >>= \case
+        True -> trace $ "Waiting for full resolution of " ++ condense pid ++ "'s tick..."
+        False -> trace $ "Client for " ++ condense pid ++ " has ceased operation."
+      threadDelay 0.100
+      trace $ condense pid ++ "'s tick is now done."
+    Nothing -> error "“The impossible happened,” as GHC would say."
+  where
+    trace = traceUnitWith tracer "Scheduler"
+
+-- | Iterate over a 'PointSchedule', sending each tick to the associated peer in turn,
+-- giving each peer a chunk of computation time, sequentially, until it satisfies the
+-- conditions given by the tick.
+-- This usually means for the ChainSync server to have sent the target header to the
+-- client.
+runScheduler ::
+  IOLike m =>
+  Tracer m String ->
+  PointSchedule ->
+  Map PeerId (PeerResources m) ->
+  m ()
+runScheduler tracer (PointSchedule ps) peers = do
+  traceWith tracer "Schedule is:"
+  for_ ps  $ \tick -> traceWith tracer $ "  " ++ condense tick
+  traceWith tracer "--------------------------------------------------------------------------------"
+  traceWith tracer "» Time says “Let there be”"
```

Comment:
Perhaps we should replace this with something that describes what these messages are delimiting.